### DISCUSSION TITLE
Feature Request: API for updating shared layouts

### DISCUSSION THREAD

**Discussion by rg-gravis** [foxglove/community#1054] [Q&A]
Use case: we have ~6 organization layouts with minor differences that we generate `.json` files for programmatically, based on a single template. It's currently a pain to update the shared layouts after regenerating the json: you need to upload all the new ones as personal layouts, delete all the old shared layouts, and then share all the new layouts one-by-one. 

It would be better if the script which generates the layouts could also push them to the organization. A PATCH endpoint in the foxglove API that allows me update a shared layout by specifying its layout ID and providing a new layout json would meet my needs here.

**Comment by banisadr** [comment]
Thanks for the feedback @rg-gravis. The reason we don't have an API for this at the moment is the layout `.json` is not a stable API. We're constantly adding and updating features, so making it a stable, versioned, non-breaking API would slow things at our current state.

We are considering this for the future. We're also looking into other options that solve the primary reasons we hear folks asking for this type of thing. Can you share more details on what you're doing? What's the purpose/functionality you get out of having a template layout and generating variations on it (seemingly regularly)?

**Reply by rg-gravis** [reply]
Sure! Our product is an automation kit for controlling existing machines, not a machine itself, so different machines equipped with the same product have different e.g. numbers of joints. Currently our main use for these different layouts is that we want a plot of state, error, etc. for each joint, so we need to add or remove plots from the layout depending on which machine we're currently visualizing (we have a very dense layout so we don't want to waste space on a plot with no data). But for everything in our layouts not related to joints, we want all the layouts to stay the same as each other, because it's fundamentally the same product. Therefore, when we want to add or change non-joint-related panels we modify a single "template" layout with the maximum number of joints, and then the script generates a layout per machine by removing all the excess plots (and adjusting the `splitPercentage`)
We also assign joint indices differently per robot, and for that we have the indices set as global variables and each per-machine layout has different values for those variables. There's 5 variables so being able to switch them all, simultaneously, via the layout selector is much faster than changing them all manually.
